src/communication/communication_channel.cpp

```diff
@@ -0,0 +1,297 @@
+}
+
+int communication_channel::send (const std::string &message)
+{
+  return communication_channel::send (message.c_str (), message.length ());
+}
+
+int communication_channel::recv (char *buffer, int &received_length)
+{
+  return css_net_recv (m_socket, buffer, &received_length, m_max_timeout_in_ms);
+}
+
+int communication_channel::connect ()
+{
+  int length = 0;
+  int rc = NO_ERRORS;
+
+  assert (m_type == CHANNEL_TYPE::INITIATOR);
+  assert (m_command_type > NULL_REQUEST && m_command_type < MAX_REQUEST);
+
+  if (m_type != CHANNEL_TYPE::INITIATOR || is_connection_alive ())
+    {
+      return ER_FAILED;
+    }
+
+  if (m_server_name == nullptr)
+    {
+      length = 0;
+    }
+  else
+    {
+      length = strlen (m_server_name.get ());
+    }
+
+  m_socket = css_tcp_client_open (m_hostname.get (), m_port);
+  if (!IS_INVALID_SOCKET (m_socket))
+    {
+      /* send magic */
+      NET_HEADER magic_header = DEFAULT_HEADER_DATA, command_header = DEFAULT_HEADER_DATA, data_header = DEFAULT_HEADER_DATA;
+      memset ((char *) &magic_header, 0, sizeof (NET_HEADER));
+      memcpy ((char *) &magic_header, css_Net_magic, sizeof (css_Net_magic));
+
+      rc = communication_channel::send ((const char *) &magic_header, sizeof (NET_HEADER));
+      if (rc != NO_ERRORS)
+	{
+	  close_connection ();
+	  return REQUEST_REFUSED;
+	}
+
+      /* send request */
+      css_set_net_header (&command_header, COMMAND_TYPE, m_command_type, m_request_id, length, 0, 0, 0);
+      if (length > 0)
+	{
+	  css_set_net_header (&data_header, DATA_TYPE, 0, m_request_id, length, 0, 0, 0);
+	  rc = communication_channel::send ((char *) &command_header, sizeof (NET_HEADER), (char *) &data_header,
+					    sizeof (NET_HEADER), m_server_name.get (), length);
+	  if (rc != NO_ERRORS)
+	    {
+	      close_connection ();
+	      return REQUEST_REFUSED;
+	    }
+	}
+      else
+	{
+	  rc = communication_channel::send ((char *) &command_header, sizeof (NET_HEADER));
+	  if (rc != NO_ERRORS)
+	    {
+	      close_connection ();
+	      return REQUEST_REFUSED;
+	    }
+	}
+
+      _er_log_debug (ARG_FILE_LINE, "connect_to_master:" "connected to master_hostname:%s\n", m_hostname.get ());
+      return NO_ERRORS;
+    }
+  else
+    {
+      return REQUEST_REFUSED;
+    }
+}
+
+int communication_channel::connect (const char *hostname, int port, CSS_COMMAND_TYPE command_type,
+				    const char *server_name)
+{
+  if (is_connection_alive ())
+    {
+      return ER_FAILED;
+    }
+
+  char *old_hostname = m_hostname.release ();
+  char *old_server_name = m_server_name.release ();
+
+  free (old_hostname);
+  free (old_server_name);
+
+  m_hostname = std::unique_ptr <char, communication_channel_c_free_deleter> (hostname == NULL? nullptr : strdup (
+		 hostname));
+  m_server_name = std::unique_ptr <char, communication_channel_c_free_deleter> (server_name == NULL ? nullptr : strdup (
+		    server_name));
+  m_port = port;
+  m_command_type = command_type;
```

Comment:
I think we will always use DATA_REQUEST for command_type. Do you think we need custom command_type?

Reply:
couldn't a communication_channel connect to something and only ask for info? (this is an INFO_REQUEST)

Reply:
I don't think communication_channel will be used for this purpose. It should be created to send and receive messages.

Alternatively, if such a communication_channel is required, it may be a different specialization.